File: src/main/java/com/databricks/labs/guidewire/ParquetStream.java
package com.databricks.labs.guidewire;

import org.apache.parquet.io.DelegatingSeekableInputStream;
import org.apache.parquet.io.InputFile;
import org.apache.parquet.io.SeekableInputStream;

import java.io.ByteArrayInputStream;

public class ParquetStream implements InputFile {

    private final byte[] data;

    public ParquetStream(byte[] data) {
        this.data = data;
    }

    @Override
    public long getLength() {
        return this.data.length;
    }

    @Override
    public SeekableInputStream newStream() {
        return new DelegatingSeekableInputStream(new SeekableByteArrayInputStream(this.data)) {
            @Override
            public void seek(long newPos) {
                ((SeekableByteArrayInputStream) this.getStream()).setPos((int) newPos);
            }

            @Override
            public long getPos() {
                return ((SeekableByteArrayInputStream) this.getStream()).getPos();
            }
        };
    }

    private static class SeekableByteArrayInputStream extends ByteArrayInputStream {
        public SeekableByteArrayInputStream(byte[] buf) {
            super(buf);
        }

        public int getPos() {
            return this.pos;
        }

        public void setPos(int pos) {
            this.pos = pos;
        }
    }
}

File: src/main/scala/com/databricks/labs/guidewire/ParquetStream.scala
package com.databricks.labs.guidewire

import org.apache.parquet.io.{DelegatingSeekableInputStream, InputFile, SeekableInputStream}

import java.io.ByteArrayInputStream

class ParquetStream(data: Array[Byte]) extends InputFile {

  override def getLength: Long = data.length

  override def newStream(): SeekableInputStream = {
    new DelegatingSeekableInputStream(new SeekableByteArrayInputStream(this.data)) {
      override def seek(newPos: Long): Unit =
        this.getStream.asInstanceOf[SeekableByteArrayInputStream].setPos(newPos.toInt)

      override def getPos: Long =
        this.getStream.asInstanceOf[SeekableByteArrayInputStream].getPos
    }
  }

  private class SeekableByteArrayInputStream(buf: Array[Byte]) extends ByteArrayInputStream(buf) {
    def getPos: Int = this.pos

    def setPos(pos: Int): Unit = this.pos = pos
  }
}
